modin/backends/pandas/query_compiler.py

```diff
@@ -2576,34 +2576,71 @@ def groupby_agg(
         if callable(agg_func):
             agg_func = wrap_udf_function(agg_func)
 
-        if is_multi_by:
-            return super().groupby_agg(
-                by=by,
-                is_multi_by=is_multi_by,
-                axis=axis,
-                agg_func=agg_func,
-                agg_args=agg_args,
-                agg_kwargs=agg_kwargs,
-                groupby_kwargs=groupby_kwargs,
-                drop=drop,
-            )
-
-        by = by.to_pandas().squeeze() if isinstance(by, type(self)) else by
-
         # since we're going to modify `groupby_kwargs` dict in a `groupby_agg_builder`,
         # we want to copy it to not propagate these changes into source dict, in case
         # of unsuccessful end of function
         groupby_kwargs = groupby_kwargs.copy()
 
         as_index = groupby_kwargs.get("as_index", True)
 
-        def groupby_agg_builder(df):
+        if isinstance(by, type(self)):
+            internal_by = by.columns
+            by = [by]
+        else:
+            if not isinstance(by, list):
+                by = [by]
+            internal_by = [o for o in by if isinstance(o, str) and o in self.columns]
+            internal_qc = (
+                [self.getitem_column_array(internal_by)] if len(internal_by) else []
+            )
+
+            by = internal_qc + by[len(internal_by) :]
+
+        broadcastable_by = [o._modin_frame for o in by if isinstance(o, type(self))]
```

Comment:
What is the difference between `broadcastable_by ` and `not_broadcastable_by `?

Reply:
`broadcast_apply_full_axis` broadcasts only modin frames, so we have to split 'by's into two different classes:
1. `broadcastable_by` - 'by' that has modin frame and will be broadcasted via `broadcast_apply_full_axis`
2. `not_broadcastable_by` - all other types of 'by' such as lists of values, names of index levels, etc. These 'by' will be captured by `groupby_agg_builder` via closure

Reply:
Got it.